File: packages/gbnf-experiment/data/20260908T234247Z_2da5a2c0/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
"""Port of ``src/grammar-graph/get-parent-stack-id.ts``."""

from __future__ import annotations

from typing import TYPE_CHECKING, Callable, List, Union

from .colorize import Color

if TYPE_CHECKING:  # pragma: no cover
    from .graph_pointer import GraphPointer

Colorize = Callable[[Union[str, int], Color], str]


def get_parent_stack_id(pointer: "GraphPointer", col: Colorize) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent is not None:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(id, Color.RED) for id in stack_ids)

File: proposed-projects/round-trip-experiment/reverse/20260909T154334Z_609edff1/20260910T091410Z_185eaa51/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color, Colorize } from './colorize';
import type { GraphPointer } from './graph-pointer';

export const getParentStackId = (pointer: GraphPointer, col: Colorize): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent) {
    const { stackId, pathId, stepId } = parent.node.meta;
    stackIds.push(`${stackId},${pathId},${stepId}`);
    parent = parent.parent;
  }
  return stackIds.map(id => col(id, Color.RED)).join(col('<-', Color.GRAY));
};
